src/percy-agent-client/dom.ts

```diff
@@ -0,0 +1,205 @@
+/**
+ * A single class to encapsulate all DOM operations that need to be preformed to
+ * capture the customers application state.
+ *
+ */
+class DOM {
+  originalDOM: any
+  clonedDOM: any
+  options: any
+
+  readonly defaultDoctype = '<!DOCTYPE html>'
+
+  constructor(dom: any, options?: object) {
+    this.originalDOM = dom
+    this.options = options || {}
+    this.clonedDOM = this.cloneDOM()
+  }
+
+  /**
+   * Returns the final DOM string with all of the necessary transforms
+   * applied. This is the string that is passed to the API and then rendered by
+   * our API.
+   *
+   */
+  snapshotString(): string {
+    let dom = this.clonedDOM
+    const doctype = this.getDoctype()
+
+    // Sometimes you'll want to transform the DOM provided into one ready for snapshotting
+    // For example, if your test suite runs tests in an element inside a page that
+    // lists all yours tests. You'll want to "hoist" the contents of the testing container to be
+    // the full page. Using a dom transformation is how you'd acheive that.
+    if (this.options.domTransformation) {
+      try {
+        dom = this.options.domTransformation(dom)
+      } catch (error) {
+        console.error('Could not transform the dom: ', error)
+      }
+    }
+
+    const snapshotString = doctype + dom.outerHTML
```

Comment:
Probably an unneeded var here

Reply:
Yeah agreed. Totally fine to just shove this into the return value directly.